cpp/oneapi/dal/table/detail/csr.hpp

```diff
@@ -0,0 +1,196 @@
+    /// performed that this type is the actual type of the data within the table.
+    template <typename Data>
+    const Data* get_data() const {
+        return reinterpret_cast<const Data*>(this->get_data());
+    }
+
+    /// The pointer to the data block within the table.
+    /// Should be equal to ``nullptr`` when :expr:`row_count == 0` and :expr:`column_count == 0`.
+    const void* get_data() const;
+
+    /// Returns the :literal:`column_indices` pointer.
+    const std::int64_t* get_column_indices() const;
+
+    /// Returns the :literal:`row_indices` pointer.
+    const std::int64_t* get_row_indices() const;
+
+private:
+    explicit csr_table(detail::csr_table_iface* impl) : table(impl) {}
+
+    bool correct_indices(const std::int64_t size,
+                         const std::int64_t* indices,
+                         const csr_indexing indexing) const {
+        const std::int64_t min_value = (indexing == csr_indexing::zero_based) ? 0 : 1;
+        for (std::int64_t i = 0; i < size; i++) {
+            if (indices[i] < min_value) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    template <typename Policy,
+              typename Data,
+              typename DataDeleter,
+              typename ColumnDeleter,
+              typename RowDeleter>
+    void init_impl(const Policy& policy,
+                   std::int64_t column_count,
+                   std::int64_t row_count,
+                   const Data* data,
+                   const std::int64_t* column_indices,
+                   const std::int64_t* row_indices,
+                   DataDeleter&& data_deleter,
+                   ColumnDeleter&& column_deleter,
+                   RowDeleter&& row_deleter,
+                   csr_indexing indexing) {
+        using error_msg = dal::detail::error_messages;
+
+        if (row_count <= 0) {
+            throw dal::domain_error(error_msg::rc_leq_zero());
+        }
+
+        if (column_count <= 0) {
+            throw dal::domain_error(error_msg::cc_leq_zero());
+        }
+
+        if (!correct_indices(column_count, column_indices, indexing) ||
+            !correct_indices(row_count, row_indices, indexing)) {
+            throw dal::domain_error(error_msg::invalid_indices());
+        }
+
+        array<Data> data_array{ data,
+                                row_indices[row_count] - row_indices[0],
+                                std::forward<DataDeleter>(data_deleter) };
+
+        array<std::int64_t> column_indices_array{ column_indices,
+                                                  row_indices[row_count] - row_indices[0],
+                                                  std::forward<ColumnDeleter>(column_deleter) };
+
+        array<std::int64_t> row_indices_array{ row_indices,
+                                               row_count + 1,
+                                               std::forward<RowDeleter>(row_deleter) };
```

Comment:
If you are passing policy, please use it to properly allocate the data. Leave USM branch unimplemented and throw an exception from it

Reply:
I suggest to discuss about it offline